In [7]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

df = pd.read_csv('Final_Imputed_Whole.csv',index_col=0)
print(df)


      population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.19           0.33          0.02          0.90          0.12   
1           0.00           0.16          0.12          0.74          0.45   
2           0.00           0.42          0.49          0.56          0.17   
3           0.04           0.77          1.00          0.08          0.12   
4           0.01           0.55          0.02          0.95          0.09   
...          ...            ...           ...           ...           ...   
1989        0.01           0.40          0.10          0.87          0.12   
1990        0.05           0.96          0.46          0.28          0.83   
1991        0.16           0.37          0.25          0.69          0.04   
1992        0.08           0.51          0.06          0.87          0.22   
1993        0.20           0.78          0.14          0.46          0.24   

      racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  ...  

In [8]:
df_train = df.iloc[0:1495,:]
df_test = df.iloc[1495:1994,:]
print(df_train)
print(df_test)

      population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.19           0.33          0.02          0.90          0.12   
1           0.00           0.16          0.12          0.74          0.45   
2           0.00           0.42          0.49          0.56          0.17   
3           0.04           0.77          1.00          0.08          0.12   
4           0.01           0.55          0.02          0.95          0.09   
...          ...            ...           ...           ...           ...   
1490        0.00           0.59          0.01          0.69          1.00   
1491        0.16           0.62          0.03          0.74          0.48   
1492        0.01           0.36          0.95          0.24          0.03   
1493        0.06           0.53          0.01          0.98          0.05   
1494        0.05           0.43          0.08          0.65          0.12   

      racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  ...  

In [9]:
df_test.reset_index(drop=True,inplace=True)
print(df_test)

     population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0          0.00           0.20          0.03          0.91          0.23   
1          0.00           0.55          0.13          0.78          0.15   
2          0.00           0.52          0.04          0.59          0.05   
3          0.04           0.51          0.07          0.92          0.09   
4          0.03           0.59          0.20          0.75          0.19   
..          ...            ...           ...           ...           ...   
494        0.01           0.40          0.10          0.87          0.12   
495        0.05           0.96          0.46          0.28          0.83   
496        0.16           0.37          0.25          0.69          0.04   
497        0.08           0.51          0.06          0.87          0.22   
498        0.20           0.78          0.14          0.46          0.24   

     racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  ...  \
0          

In [10]:
X_train = df_train.drop(columns=['ViolentCrimesPerPop'])
X_test = df_test.drop(columns=['ViolentCrimesPerPop'])
y_train = df_train['ViolentCrimesPerPop']
y_test = df_test['ViolentCrimesPerPop']
print(X_train)
print(X_test)
print(y_train)
print(y_test)


      population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.19           0.33          0.02          0.90          0.12   
1           0.00           0.16          0.12          0.74          0.45   
2           0.00           0.42          0.49          0.56          0.17   
3           0.04           0.77          1.00          0.08          0.12   
4           0.01           0.55          0.02          0.95          0.09   
...          ...            ...           ...           ...           ...   
1490        0.00           0.59          0.01          0.69          1.00   
1491        0.16           0.62          0.03          0.74          0.48   
1492        0.01           0.36          0.95          0.24          0.03   
1493        0.06           0.53          0.01          0.98          0.05   
1494        0.05           0.43          0.08          0.65          0.12   

      racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  ...  

In [11]:
reg = LinearRegression().fit(X_train,y_train)
y_predicted = reg.predict(X_test)
print(y_test)
print(y_predicted)

0      0.12
1      0.11
2      0.12
3      0.06
4      0.12
       ... 
494    0.09
495    0.45
496    0.23
497    0.19
498    0.48
Name: ViolentCrimesPerPop, Length: 499, dtype: float64
[ 1.15996343e-01 -2.72448756e-02  1.48671641e-01  9.04148809e-03
  1.40651125e-01  9.15078590e-03  2.28797374e-01  4.71143133e-01
  1.30761345e-01  3.48767590e-01  2.79212266e-01  5.83856250e-01
  2.49861501e-01  5.73066038e-01  4.12491332e-01  1.80280459e-01
  3.03612793e-01  4.56541054e-01  2.24987941e-01  4.11679732e-01
  1.69747066e-02  4.01617534e-02  2.70944224e-01  2.49455907e-01
  2.55072758e-03  3.13290909e-02  1.04544413e-01  1.54563259e-02
  4.62376931e-01  6.99695614e-01  1.37372720e-01  1.99949602e-01
  2.22228618e-01  2.22858058e-01  1.08576647e-01  6.37167138e-02
  6.10657828e-02  2.71406159e-01  1.75974059e-01  4.08664899e-01
  1.77243429e-01  6.74366027e-01  2.92167752e-01  4.46202206e-01
  4.14699555e-01  3.32588533e-01  3.47977569e-01  2.70340322e-01
  6.74872104e-01  5.07380730e-01 

Due to imputation we do get a minor amount of values negative which are extremely close to zero. Negative values, as such, are impossible for our outcome in general.

In [12]:
MSE = mean_squared_error(y_test,y_predicted)
print(MSE)

0.056641183992459446


The test MSE thus obtained is 0.0566. As such, while this value may seem rather small, it is still relatively noticeable given that our data has been normalized.

Square root of MSE, gives, on average, deviation (absolute) of the predicted value from the true value, in this case about 0.23, which is quite excessive given that output lies in the range of (0,1) (normalized)